### PR TITLE
MAE-751: Consolidate logic for calculating membership instalment dates

### PR DESCRIPTION
**Pull request by olayiwola-compucorp**
## Overview
The instalment dates listed on the new membership screen can sometimes not match the contribution instalment dates after creation, this difference is noticeable when using a day above 28. In this PR we use the same logic used in calculating contribution dates to calculate the instalment dates for the new membership screen.

## Before
The instalment dates on the new membership screen.
Screenshot 2022-06-15 at 09 04 37

The new membership screen uses the inbuilt PHP DateTime function to add months to date, and this function behaves in some certain way as we've seen in the screen above, `2022-01-29 + 1month =  2022-03-01` and subsequent months use the day from the previous calculation. 

The actual instalment dates after creation.
date

Whereas when creating the contributions, the instalment dates are calculated using the custom function `getSameDayNextMonth` https://github.com/compucorp/uk.co.compucorp.membershipextras/blob/326e69ed0468cb4f2b40ad6adc0b6630257e84d3/CRM/MembershipExtras/Service/InstalmentReceiveDateCalculator.php#L102 to add months to date, this function helps handle the scenario mentioned above such that `2022-01-29 + 1month =  2022-02-28` and subsequent dates makes use of the day from the first calculation. 

## After
New membership screen using the function `getSameDayNextMonth` to calculate instalment dates
Screenshot 2022-06-15 at 09 31 48
Screenshot 2022-06-15 at 09 32 02
Screenshot 2022-06-15 at 09 32 13

### CODE REVIEW

**Review by omarabuhussein — CHANGES_REQUESTED** [review]
Can you add tests to CRM_MembershipExtras_Service_MembershipInstalmentsScheduleTest to cover these  cover such edge cases.

File: tests/phpunit/CRM/MembershipExtras/Service/MembershipInstalmentsScheduleTest.php [hunk 1]
```diff
@@ -644,6 +644,30 @@ public function testExceptionIsThrownIfMembershipTypeDurationUnitIsDay() {
     $this->getMembershipInstalmentsSchedule([$membershipType], MembershipInstalmentsSchedule::ANNUAL);
   }
 
+  /**
+   * Tests that instalments date are calculated properly and months follows a sequential order.
+   */
+  public function testRollingInstalmentDatesAreCalculatedProperly() {
```
Comment (omarabuhussein) [inline tests/phpunit/CRM/MembershipExtras/Service/MembershipInstalmentsScheduleTest.php:650]: the name of the test should reflect the edge case that you are testing (which is that start date day is > 28).

File: tests/phpunit/CRM/MembershipExtras/Service/MembershipInstalmentsScheduleTest.php [hunk 2]
```diff
@@ -644,6 +644,30 @@ public function testExceptionIsThrownIfMembershipTypeDurationUnitIsDay() {
...
+    );
+
+    foreach ($schedule['instalments'] as $index => $instalment) {
+      $instalmentDate = $instalment->getInstalmentDate();
+      $year = $instalmentDate->format('Y');
+      $month = $instalmentDate->format('n');
+      $day = $instalmentDate->format('d');
+
+      $this->assertTrue(checkdate($month, $day, $year));
+      $this->assertEquals($month, $currentMonth++);
```
Comment (omarabuhussein) [inline tests/phpunit/CRM/MembershipExtras/Service/MembershipInstalmentsScheduleTest.php:667]: testing the month is correct is not enough to make sure the date is calculated correctly, a more accurate test should include an array that include all the expected installment dates (all hardcoded), and you compare the resulted dates against them.

**Review by omarabuhussein — CHANGES_REQUESTED** [review]

File: tests/phpunit/CRM/MembershipExtras/Service/MembershipInstalmentsScheduleTest.php (hunk 2 above)
Comment (olayiwola-compucorp) [reply tests/phpunit/CRM/MembershipExtras/Service/MembershipInstalmentsScheduleTest.php:667]: Thanks, @omarabuhussein.

> a more accurate test should include an array that include all the expected installment dates (all hardcoded)

I avoided doing this, since we have tests like `testRollingMonthlyInstalmentAmounts` and others in place, and  I thought they were enough to assert the correctness of the returned date.

e.g. if the `start_date` is 2022-01-31 and the next date is 2022-03-01 the test will fail.

**Review by omarabuhussein — APPROVED** [review]
